File: apps/api/app/schemas/annotations.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.annotation import Annotation, AnnotationType


class AnnotationCreate(BaseModel):
    type: str = AnnotationType.HIGHLIGHT
    quote: str = ""
    start_offset: int | None = None
    end_offset: int | None = None
    page: int | None = None
    color: str | None = None
    body: str | None = None
    labels: list[str] = Field(default_factory=list)
    pdf_anchor: dict[str, Any] | None = None


class AnnotationUpdate(BaseModel):
    type: str | None = None
    quote: str | None = None
    start_offset: int | None = None
    end_offset: int | None = None
    page: int | None = None
    color: str | None = None
    body: str | None = None
    labels: list[str] | None = None
    pdf_anchor: dict[str, Any] | None = None


class AnnotationRead(BaseModel):
    id: str
    document_id: str
    type: str
    quote: str
    start_offset: int | None
    end_offset: int | None
    page: int | None
    color: str | None
    body: str | None
    labels: list[str]
    pdf_anchor: dict[str, Any] | None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_model(cls, ann: Annotation) -> AnnotationRead:
        return cls(
            id=ann.id,
            document_id=ann.document_id,
            type=ann.type,
            quote=ann.quote,
            start_offset=ann.start_offset,
            end_offset=ann.end_offset,
            page=ann.page,
            color=ann.color,
            body=ann.body,
            labels=ann.labels or [],
            pdf_anchor=ann.pdf_anchor,
            created_at=ann.created_at,
            updated_at=ann.updated_at,
        )

File: apps/web-lite/src/local-api/annotations.ts
// Straight IndexedDB CRUD — the backend never computed anything for annotations, it only stored
// them, so this is a drop-in for apps/web/lib/api/annotations.ts.
import type { Annotation, AnnotationCreate } from '@web/lib/types';
import { getDb, newId, utcnow } from '../db/db';
import { notFound } from './errors';

export async function listAnnotations(documentId: string): Promise<Annotation[]> {
  const db = await getDb();
  const items = await db.getAllFromIndex('annotations', 'by_document', documentId);
  items.sort((a, b) => (a.created_at < b.created_at ? -1 : 1));
  return items;
}

export async function createAnnotation(
  documentId: string,
  payload: AnnotationCreate,
): Promise<Annotation> {
  const now = utcnow();
  const annotation: Annotation = {
    id: newId(),
    document_id: documentId,
    type: payload.type,
    quote: payload.quote,
    start_offset: payload.start_offset ?? null,
    end_offset: payload.end_offset ?? null,
    page: payload.page ?? null,
    color: payload.color ?? null,
    body: payload.body ?? null,
    labels: payload.labels ?? [],
    pdf_anchor: payload.pdf_anchor ?? null,
    created_at: now,
    updated_at: now,
  };
  const db = await getDb();
  await db.put('annotations', annotation);
  return annotation;
}

export async function updateAnnotation(
  annotationId: string,
  payload: Partial<AnnotationCreate>,
): Promise<Annotation> {
  const db = await getDb();
  const existing = await db.get('annotations', annotationId);
  if (!existing) throw notFound('Annotation');
  const updated: Annotation = {
    ...existing,
    ...(payload.type !== undefined ? { type: payload.type } : {}),
    ...(payload.quote !== undefined ? { quote: payload.quote } : {}),
    ...(payload.start_offset !== undefined ? { start_offset: payload.start_offset } : {}),
    ...(payload.end_offset !== undefined ? { end_offset: payload.end_offset } : {}),
    ...(payload.page !== undefined ? { page: payload.page } : {}),
    ...(payload.color !== undefined ? { color: payload.color } : {}),
    ...(payload.body !== undefined ? { body: payload.body } : {}),
    ...(payload.labels !== undefined ? { labels: payload.labels ?? [] } : {}),
    ...(payload.pdf_anchor !== undefined ? { pdf_anchor: payload.pdf_anchor } : {}),
    updated_at: utcnow(),
  };
  await db.put('annotations', updated);
  return updated;
}

export async function deleteAnnotation(annotationId: string): Promise<void> {
  const db = await getDb();
  await db.delete('annotations', annotationId);
}
